### PR TITLE
Update NonEmptyLeaseTableSynchronizer algorithm to create leases for no more than one level of the shard hierarchy

### PR DESCRIPTION
**Pull request by micah-jaffe**
Modifying the memoization between recursive calls to only process a single layer of the shard hierarchy at a time. All existing and new unit tests pass. Had to rejigger a lot of the existing unit tests since the behavior of this lease synchronizer is changing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 1]
```diff
@@ -1095,45 +1151,29 @@ public int compare(final Lease lease1, final Lease lease2) {
                 } else if (inconsistentShardIds.contains(shardId)) {
                     log.info("{} : shardId {} is an inconsistent child.  Not creating a lease", streamIdentifier, shardId);
                 } else {
-                    log.debug("{} : Need to create a lease for shardId {}", streamIdentifier, shardId);
-                    final Lease newLease = multiStreamArgs.isMultiStreamMode() ?
-                            newKCLMultiStreamLease(shard, multiStreamArgs.streamIdentifier()) :
-                            newKCLLease(shard);
+                    log.debug("{} : Need to create a lease in ancestry tree for shardId {}", streamIdentifier, shardId);
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Need to create ancestor lease will not be conclusive at this point

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 2]
```diff
@@ -1095,45 +1151,29 @@ public int compare(final Lease lease1, final Lease lease2) {
...
+                    // SHARD_END of their parents.
+                    if (!isDescendant) {
+                        log.debug("{} : shardId {} has no ancestors. Creating a lease.", streamIdentifier, shardId);
+                        final Lease newLease = multiStreamArgs.isMultiStreamMode() ?
+                                newKCLMultiStreamLease(shard, multiStreamArgs.streamIdentifier()) :
+                                newKCLLease(shard);
                         newLease.checkpoint(convertToCheckpoint(initialPosition));
+                        log.debug("{} : Set checkpoint of {} to {}", streamIdentifier, newLease.leaseKey(), newLease.checkpoint());
+                        shardIdToNewLeaseMap.put(shardId, newLease);
                     }
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:1178]: can we add a debug log for else case of being a descendant ?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 3]
```diff
@@ -483,11 +485,15 @@ static boolean checkIfDescendantAndAddNewLeasesForAncestors(final String shardId
             } else {
                 final Shard shard = shardIdToShardMapOfAllKinesisShards.get(shardId);
                 final Set<String> parentShardIds = getParentShardIds(shard, shardIdToShardMapOfAllKinesisShards);
+
                 for (String parentShardId : parentShardIds) {
-                    // Check if the parent is a descendant, and include its ancestors.
+                    // Check if the parent is a descendant, and include its ancestors. Or, if the parent is NOT a
+                    // descendant but should we should create a lease for it anyway (e.g. to include in processing from
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: nit : remove extra should

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 4]
```diff
@@ -474,13 +476,19 @@ public void testCheckAndCreateLeasesForShardsWithEmptyShardList() throws Excepti
     }
 
     @Test
-    public void testCheckAndCreateLeasesForNewShardsAtTrimHorizon() throws Exception {
-        testCheckAndCreateLeaseForShardsIfMissing(constructShardListForGraphA(), INITIAL_POSITION_TRIM_HORIZON);
+    public void testCheckAndCreateLeasesForNewShardsAtTrimHorizonWithPartialLeaseTable() throws Exception {
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: This is a case of empty lease table. Can we add partial lease table test cases?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 5]
```diff
@@ -474,13 +476,19 @@ public void testCheckAndCreateLeasesForShardsWithEmptyShardList() throws Excepti
...
+        final List<Shard> shards = constructShardListForGraphA();
+        final Set<String> expectedLeaseKeys = new HashSet<>(Arrays.asList("shardId-0", "shardId-1", "shardId-2",
+                "shardId-3", "shardId-4", "shardId-5"));
+        testCheckAndCreateLeaseForShardsIfMissing(shards, INITIAL_POSITION_TRIM_HORIZON, expectedLeaseKeys);
     }
 
     @Test
-    public void testCheckAndCreateLeasesForNewShardsAtTimestamp() throws Exception {
-        testCheckAndCreateLeaseForShardsIfMissing(constructShardListForGraphA(), INITIAL_POSITION_AT_TIMESTAMP);
+    public void testCheckAndCreateLeasesForNewShardsAtTimestampWithPartialLeaseTable() throws Exception {
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Same comment as above

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 6]
```diff
@@ -1342,16 +1359,20 @@ public void testDetermineNewLeasesToCreateSplitMergeHorizon2() {
         assertThat(checkpoint, equalTo(expectedCheckpoint));
     }
 
-//    /**
+    //    /**
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: nit : whitespace can be removed

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 7]
```diff
@@ -1368,6 +1389,38 @@ public void testDetermineNewLeasesToCreateGraphBNoInitialLeasesTrim() {
...
+//     */
+    @Test
+    public void testDetermineNewLeasesToCreateGraphBNoInitialLeasesTrim() {
+        final List<Shard> shards = constructShardListForGraphB();
+        final List<Lease> currentLeases = new ArrayList<>();
+
+        final HierarchicalShardSyncer.LeaseSynchronizer emptyLeaseTableSynchronizer =
+                new HierarchicalShardSyncer.EmptyLeaseTableSynchronizer();
+
+        final List<Lease> newLeases = HierarchicalShardSyncer.determineNewLeasesToCreate(emptyLeaseTableSynchronizer,
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Basically we can use the partialLTSynchronizer here as well, to see if this covers the empty lease table case as well. We know that empty lease table is simple case and can work.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 8]
```diff
@@ -1474,10 +1525,10 @@ public void testDetermineNewLeasesToCreateGraphBNoInitialLeasesAtTimestamp() {
         final Map<String, Set<String>> shardIdToChildShardIdsMap = HierarchicalShardSyncer
                 .constructShardIdToChildShardIdsMap(shardIdToShardMap);
 
-        final HierarchicalShardSyncer.LeaseSynchronizer nonEmptyLeaseTableSynchronizer =
-                new HierarchicalShardSyncer.NonEmptyLeaseTableSynchronizer(shardDetector, shardIdToShardMap, shardIdToChildShardIdsMap);
+        final HierarchicalShardSyncer.LeaseSynchronizer emptyLeaseTableSynchronizer =
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Same comment as above. Lets use partial or nonempty sync here

**Review by ashwing — COMMENTED** [review]
Overall looks good. We need more test cases to comprehensively cover all edge cases for LATEST, TRIM, AT_TS with different hierarchical graphs.  Also can you address the comments for existing test cases. Thanks for working on this!

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 1 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: True, will update

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 2 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:1178]: Sure, will add

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 4 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Good catch, will add partials to these

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 6 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Removed

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 3 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Removed

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 7 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: We can do that. This test is a bit confusing because for the empty lease table, we still want to create leases for all shards returned from the service. ListShardsWithFilter(TRIM_HORIZON) will return only shard-0 and shard-1, but this test doesn't know about the shard filter. So the expected leases from this test's perspective for an empty lease table would be shard-0, shard-1, ... shard-9, shard-10. For non-empty lease table, we use the single-level optimization so the expected leases would only be shard-0 and shard-1, which matches with the empty lease table case using ListShardsWithFilter. I'll update the test.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 9]
```diff
@@ -473,14 +474,180 @@ public void testCheckAndCreateLeasesForShardsWithEmptyShardList() throws Excepti
...
+    /*
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Missing leases: (0, 6, 8)
+     * Initial position: LATEST
+     * Leases to create: (0)
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:637]: Can you add the following cases as well? Ignore the cases already covered.
LATEST
1. If the missing leases are 0,1,6,8 or 0,1,6,7,8, then we  need to create lease only for 6
2. If the missing leases are 0,1,2,3,6,7 then we ~create leases for 6 and 7~ nothing needs to be done
3. If the missing leases are 1,2 or 1,2,6 or 1,2,6,7, we ~need to create leases for 0 and 3.~ don't do anything

TRIM/TIMESTAMP
1. If the missing leases are 0,1,6,8 or 0,1,6,7,8, then we need to create lease only for 0,1.
2. If the missing leases are 0,1,2,3,6,7 then ~we create leases for 0,1,2,3~ we dont do anything
3. If the missing leases are 1,2 or 1,2,6 or 1,2,6,7, ~we need to create leases for 0 and 3.~ then dont do anything

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 10]
```diff
@@ -1229,92 +1390,198 @@ public void testDetermineNewLeasesToCreateSplitMergeLatest1() {
...
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Current leases: (0, 1, 2, 3, 4, 5, 6, 7)
+     */
+    @Test
+    public void testDetermineNewLeasesToCreateSplitMergeLatestA_CompleteHashRangeWithoutGC() {
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:1509]: Nice. This would cover the rollforward cases.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 11]
```diff
@@ -1229,92 +1390,198 @@ public void testDetermineNewLeasesToCreateSplitMergeLatest1() {
...
+     * 1  4   7  10
+     *
+     * Current leases: empty set
+     */
     @Test
-    public void testDetermineNewLeasesToCreateSplitMergeHorizon2() {
-        final List<Shard> shards = constructShardListForGraphA();
-        final List<Lease> currentLeases = Arrays.asList(newLease("shardId-4"), newLease("shardId-5"),
-                newLease("shardId-7"));
+    public void testDetermineNewLeasesToCreateSplitMergeLatestB_EmptyLeaseTable() {
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:1637]: can we have trim, at_ts equivalent of this?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 12]
```diff
@@ -1324,175 +1591,452 @@ public void testDetermineNewLeasesToCreateSplitMergeHorizon2() {
...
-                .constructShardIdToChildShardIdsMap(shardIdToShardMap);
+    /**
+     * Test CheckIfDescendantAndAddNewLeasesForAncestors (initial position Latest)
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Current leases: (0, 1, 4, 9, 10)
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: correction : 0,1,4,7,9,10

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 13]
```diff
@@ -1324,175 +1591,452 @@ public void testDetermineNewLeasesToCreateSplitMergeHorizon2() {
...
+     * 0  3   6   9
+     * \ / \ / \ /
+     *  2   5   8
+     * / \ / \ / \
+     * 1  4   7  10
+     *
+     * Current leases: (3)
+     */
+    @Test
+    public void testDetermineNewLeasesToCreateSplitMergeHorizonB_PartialHashRange() {
```
Comment (ashwing) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Hmmm, for this case, I feel we need to just create only 4 for both latest and trim/ts. Else we might end up processing 0,1,2,4,5 leading to kind of our of order processing. Your thoughts?

**Review by ashwing — COMMENTED** [review]
Left adhoc comments. Thanks for covering most of the edge cases.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 11 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:1637]: Yep we have these below on lines 1814 and 2033

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 12 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Ack good catch, updated

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 9 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:637]: LATEST:
1. Creating a lease only for 6 and not for 0,1 is covered in testDetermineNewLeasesToCreateSplitMergeLatestA_PartialHashRange1()
2. Wouldn't we not create any leases in this case since we're reading LATEST and the hash range is complete with lease 8?
3. Not sure I understand - in this example we already have leases for 0 and 3?

TRIM/TIMESTAMP
1. Creating a lease for 0, 1 is covered in testDetermineNewLeasesToCreateSplitMergeHorizonA_PartialHashRange1() and equivalent TIMESTAMP test
2. Added tests for this: testDetermineNewLeasesToCreateSplitMergeHorizonA_PartialHashRange4 and TIMESTAMP version
3. Same as LATEST, we already have leases for 0 and 3?
  * Missing 1, 2 -> shouldn't this not create any leases since the entire hash range is covered by 8 (also 6 and 7)? This is the behavior consistent with testDetermineNewLeasesToCreateSplitMergeHorizonA_CompleteHashRange() - currently we don't backfill any ancestor leases beyond a completed hash range
  * Dependent on above, but: Missing 1, 2, 6 -> shouldn't this only create lease 1 to fill in the missing pair to lease 0?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 13 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Yeah agreed I think this is a bug. 0 and 1 are ancestors of 3, so if we're processing 3 that means we already processed 0 and 1 so we shouldn't create new leases for them. Will fix the test and the code to only create 4 in this case.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 14]
```diff
@@ -463,10 +465,10 @@ private static boolean hashRangeOfShardsIsComplete(@NonNull List<Shard> shards)
     static boolean checkIfDescendantAndAddNewLeasesForAncestors(final String shardId,
             final InitialPositionInStreamExtended initialPosition, final Set<String> shardIdsOfCurrentLeases,
             final Map<String, Shard> shardIdToShardMapOfAllKinesisShards,
```
Comment (satselva) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:467]: does shardIdToShardMap contains all the shards of the stream or the shards with respect to the initial position of the stream ?

Ex: Stream with 1000 shards and have 300 days of retention can contain 1 million shards. It takes approximately 1000 pages to complete list shards. Does this code avoid that 1000 pages of list shards ? 
If the KCL change is made to call ListShards @ TIMESTAMP, @ AT_TRIMHORIZON or AT_LATEST we will not have shards at multiple hierarchy

or Is this check used only in the recovery mode when there are holes detected ?

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 15]
```diff
@@ -1229,92 +1390,198 @@ public void testDetermineNewLeasesToCreateSplitMergeLatest1() {
...
-                Collections.singletonList(ExtendedSequenceNumber.TRIM_HORIZON));
+    /*
+     * Shard structure (x-axis is epochs):
+     * 0  3   6   9
+     * \ / \ / \ /
+     *  2   5   8
+     * / \ / \ / \
+     * 1  4   7  10
+     *
+     * Current leases: (5)
```
Comment (satselva) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:1591]: Good to add expected leases in the test cases for easy readability

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java [hunk 16]
```diff
@@ -1229,92 +1390,198 @@ public void testDetermineNewLeasesToCreateSplitMergeLatest1() {
...
+     * 1  4   7  10
+     *
+     * Current leases: (0, 1, 2, 3, 4, 5)
+     */
+    @Test
+    public void testDetermineNewLeasesToCreateSplitMergeLatestB_CompleteHashRangeWithoutGC() {
+        final List<Shard> shards = constructShardListForGraphB();
+        final List<String> shardIdsOfCurrentLeases = Arrays.asList("shardId-0", "shardId-1", "shardId-2", "shardId-3",
+                "shardId-4", "shardId-5");
+        final Map<String, ExtendedSequenceNumber> expectedShardIdCheckpointMap = Collections.emptyMap();
```
Comment (satselva) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: better to rename variable as expectedNoNewLeases

**Review by satselva — COMMENTED** [review]
Test case can be much more simplified with TestNG and data provider as most of the test cases are changes with the input set and easy to express them in TestNG. I would leave it to your preference if that would increase the scope. But will reduce and simplify test cases a lot.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 16 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Will rename

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 15 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:1591]: Sure, will add expected leases to the javadoc

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 14 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:467]: This check is only used in recovery mode when there are holes detected, at least in KCL 2.x. In KCL 1.x it could also be exercised during PeriodicShardSync but that will be handled in KCL 1.x impl.

**Comment by micah-jaffe** [comment]
> Test case can be much more simplified with TestNG and data provider as most of the test cases are changes with the input set and easy to express them in TestNG. I would leave it to your preference if that would increase the scope. But will reduce and simplify test cases a lot.

Yeah I tried doing this but TestNG isn't currently built into the package. Leaving as is for now to avoid scope increase, but this would definitely make these tests more maintainable in the future.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 9 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java:637]: Discussed offline, these cases are covered

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java (hunk 13 above)
Comment (micah-jaffe) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/HierarchicalShardSyncerTest.java]: Sent a fix for this in the new revision. It adds a bit of complexity to the code, but I feel this is an important edge case to handle. This could happen for any split operation where the shard to split has non-null parent shards and only one of its children has a lease. This could be fairly common as most shards in a frequently resharded stream have non-null parents.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 17]
```diff
@@ -484,19 +484,64 @@ static boolean checkIfDescendantAndAddNewLeasesForAncestors(final String shardId
                 // because we'd have done it when creating a lease for this shard.
             } else {
                 final Shard shard = shardIdToShardMapOfAllKinesisShards.get(shardId);
-                final Set<String> parentShardIds = getParentShardIds(shard, shardIdToShardMapOfAllKinesisShards);
 
+                final Set<String> childShardIds = shardIdToChildShardIdsMap.get(shardId);
+                if (childShardIds != null && childShardIds.size() == 2) {
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: For the case below, what will happen when we reach 5 though the path 11,9,5 but we have leases for 13 and 14. 
Would be end up creating lease for 10? Can we add test cases for these scenarios?

     * 0 1 2 3 4   5- shards till epoch 102
     * \ / \ / |   |
     *  6   7  4   5- shards from epoch 103 - 205
     *   \ /   |  /\
     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
     *          /\  /\
     *        11 12 13 14

Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: We wouldn't create a lease for 10 in this case because at the time shard 5 is being traversed (top of the stack), neither of its children will have leases so we won't hit this condition.

If we had a lease for 9, we still wouldn't create a lease for 10 because we would have a complete hash range. Will add test cases for these.

Comment (micah-jaffe) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Discussed offline. If we have leases for 13 and 14 but traverse the left half of the tree (from 11 -> 9 -> 5) before traversing the right half of the tree, we will have no way of knowing a priori at which level the hash range is complete. This breaks the TRIM/TIMESTAMP case.

Assume n = 0 is the root (shard 5). The child shards optimization in the previous revision only solves a special case where the hash range is partially complete at n = 1 (shard 10). For all n > 1, since we don't know what the other half of the tree looks like, we will go all the way back to the beginning and create a lease for shard-5. This could lead to out-of-order processing since we already have leases for 13 and 14, but will then be retroactively creating leases again for 5 (and then 10 when 5 reaches SHARD_END). 

However, this is current with the existing behavior when reading from TRIM/TIMESTAMP so there is no regression. However we should scope this work in a followup PR to modify the algorithm and fix this bug.

**Comment by ashwing** [comment]
LGTM. Please do rollback and rollforward testing with older (public KCL) and newer VFI. Allow the periodic shard syncer in old vfi to create the entire hierarchy tree and see how rollforward handles it. And vice versa.

**Review by ashwing — APPROVED** [review]
LGTM. Left a comment for testing.
